fraud/service.go

```diff
@@ -115,30 +147,190 @@ func (f *service) processIncoming(
 		"hash", hex.EncodeToString(extHeader.DAH.Hash()),
 		"from", from.String(),
 	)
-	msg.ValidatorData = proof
+	log.Warn("Shutting down services...")
+	return pubsub.ValidationAccept
+}
+
+func (f *service) processLocal(
+	ctx context.Context,
+	proofType ProofType,
+	hash string,
+	rawData []byte,
+) pubsub.ValidationResult {
+	f.storesLk.RLock()
+	storage := f.stores[proofType]
+	f.storesLk.RUnlock()
+	data, err := getByHash(ctx, storage, hash)
+	if err != nil {
+		// TODO @vgonkivs: should we panic here???
+		panic(fmt.Sprintf("unexpected error while fetching a proof from the storage: %s", err.Error()))
+	}
+
+	if bytes.Compare(data, rawData) == 0 {
+		log.Warnw("received fraud proof",
+			"proofType", proofType,
+			"headerHash", hash,
+		)
+		return pubsub.ValidationAccept
+	}
+	return pubsub.ValidationReject
+}
+
+func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
-	err = put(ctx, store, string(proof.HeaderHash()), msg.Data)
-	if err != nil {
-		log.Error(err)
-	}
-	log.Warn("Shutting down services...")
-	return pubsub.ValidationAccept
+
+	return getAll(ctx, store, proofType)
 }
 
-func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
+func (f *service) put(ctx context.Context, proofType ProofType, hash string, data []byte) error {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
+	return put(ctx, store, hash, data)
+}
 
-	return getAll(ctx, store, proofType)
+func (f *service) handleFraudMessageRequest(stream network.Stream) {
+	msg := new(pb.FraudMessage)
+	_, err := serde.Read(stream, msg)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Warn(err)
+		return
+	}
+
+	errg, ctx := errgroup.WithContext(context.Background())
+	mu := sync.Mutex{}
+	for _, p := range msg.RequestedProofType {
+		proofType := toProof(p)
+		errg.Go(func() error {
+			proofs, err := f.Get(ctx, proofType)
+			if err != nil {
+				if errors.Is(err, datastore.ErrNotFound) {
+					return nil
+				}
+				return err
+			}
+
+			for _, proof := range proofs {
+				bin, err := proof.MarshalBinary()
+				if err != nil {
+					log.Warn(err)
+					continue
+				}
+				resp := &pb.ProofResponse{Type: int32(proofType), Value: bin}
+				mu.Lock()
+				msg.Proofs = append(msg.Proofs, resp)
+				mu.Unlock()
+			}
+			return nil
+		})
+	}
+
+	if err = errg.Wait(); err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Error(err)
+		return
+	}
+	_, err = serde.Write(stream, msg)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (f *service) Sync(ctx context.Context) {
+	go f.syncFraudProofs(ctx)
+}
+
+func (f *service) syncFraudProofs(ctx context.Context) {
```

Comment:
I'm also wondering why this has to happen async.

Reply:
it relies on nodes connection. it will be blocked until 5 nodes will be discovered and connected